sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,216 @@
+
+    uint64_t start_cycles = ibex_mcycle_read();
+    test->func(buf1, buf2, i);
+    uint64_t end_cycles = ibex_mcycle_read();
+
+    // Practically, we should never see the 64-bit cycle counter overflow. Even
+    // if it took 12 hours to run all the perf tests, the clock would have to
+    // run at 427 THz to overflow the counter (2**64 cycles / (12 * 60 * 60)
+    // seconds).
+    CHECK(start_cycles < end_cycles);
+    const uint64_t num_cycles = end_cycles - start_cycles;
+
+    CHECK(total_clock_cycles < UINT64_MAX - num_cycles);
+    total_clock_cycles += num_cycles;
+  }
+
+  return total_clock_cycles;
+}
+
+// Fill the buffer with arbitrary, but deterministically-selected bytes.
+static inline void fill_buf_deterministic_values(uint8_t *buf, size_t len) {
+  uint32_t state = 42;
+  for (size_t i = 0; i < len; ++i) {
+    state = state * 17 + i;
+    buf[i] = (uint8_t)state;
+  }
+}
+
+// Zero out the buffer.
+static inline void fill_buf_zeroes(uint8_t *buf, size_t len) {
+  memset(buf, 0, len);
+}
+
+// Zero out the buffer, but put a one at the end.
+static inline void fill_buf_zeroes_then_one(uint8_t *buf, size_t len) {
+  fill_buf_zeroes(buf, len);
+  buf[len - 1] = 1;
+}
+
+// Zero out the buffer, but put a one at the beginning.
+static inline void fill_buf_one_then_zeroes(uint8_t *buf, size_t len) {
+  fill_buf_zeroes(buf, len);
+  buf[0] = 1;
+}
+
+OT_NEVER_INLINE void test_memcpy(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  memcpy(buf1, buf2, len);
+}
+
+OT_NEVER_INLINE void test_memset(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  const int value = buf2[0];
+  memset(buf1, value, len);
+}
+
+OT_NEVER_INLINE void test_memcmp(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  memcmp(buf1, buf2, len);
+}
+
+OT_NEVER_INLINE void test_memrcmp(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  memrcmp(buf1, buf2, len);
+}
+
+OT_NEVER_INLINE void test_memchr(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  const uint8_t value = buf1[len - 1];
+  memchr(buf1, value, len);
+}
+
+OT_NEVER_INLINE void test_memrchr(uint8_t *buf1, uint8_t *buf2, size_t len) {
+  const uint8_t value = buf1[0];
+  memrchr(buf1, value, len);
+}
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// Each value of `expected_num_cycles` was determined experimentally by
+// testing on a CW310 FPGA with the following command:
+//
+//   $ ./bazelisk.sh test --copt -O2 --test_output=all \
+//       //sw/device/lib/base:memory_perftest_fpga_cw310
+//
+// There are a handful of reasons why the expected number of cycles for this
+// test might be inaccurate. Here are a few of them:
+//
+//   (1) You've changed the definition of the test, e.g. changing the size of
+//       the test buffers.
+//   (2) You've changed the implementation of memset, memcpy, etc. and they can
+//       do the same job in fewer cycles.
+//   (3) The test was not compiled with `-O2`. The hardcoded cycle count
+//       expectations assume `-O2`.
+//   (4) The compiler has gotten smarter.
```

Comment:
(5) The icache gets turned on prior to test execution.

Reply:
Thanks!